### DISCUSSION TITLE
upip to mip migration

### DISCUSSION THREAD

**Discussion by simonmccaughey** [micropython/micropython#9565] [Libraries & Drivers] [answered]
Hi, I understand that mip replaces upip, but I can't install my normal required packages with mip - I'm wondering did they change name, or can I direct mip to a different url to pick up the old packages? I tried searching, but I can't find any migration tips - reaching out for help. I'm using the rpi pico-w board.

This is my old installation script:
```
import upip
upip.install('micropython-ulogging')
upip.install('micropython-ssd1306')
upip.install('picoweb')
```

example result:
```
>>> mip.install('picoweb')
Installing picoweb (latest) from https://micropython.org/pi/v2 to /lib
Package not found: https://micropython.org/pi/v2/package/6/picoweb/latest.json
Package may be partially installed
>>> 
```

**Comment by jimmo** [comment, marked as answer]
Hi @simonmccaughey -- thanks for raising this, it's difficult to know how people were previously using `upip`.

This actually perfectly highlights one of the things we were keen to leave behind with `mip` -- `upip` was actually quite confusing as to where the packages actually came from. Also the difference between "foo" and "micropython-foo". The idea now is that _everything_ comes from micropython-lib, unless it's via a custom URL (i.e. starts with http:// or github:), or using a custom index (via the index kwarg).

In this case:
`micropython-ulogging` -- this actually came from micropython-lib.
`micropython-ssd1306` -- this was [forked from the official driver by someone](https://github.com/stlehmann/micropython-ssd1306) just for the purpose of making it available in upip. (But it actually comes from the main micropython repo)
`picoweb` -- this comes from pycopy

Anyway, to answer your question, here's how you install the first two now:

```
>>> mip.install("logging")
>>> mip.install("ssd1306")
```

For logging,  i'm not sure why the pypi package was called ulogging, the version in micropython-lib is called "logging". I've confirmed this is definitely the same code.

For ssd1306, the official driver is now in micropython lib and therefore in `mip`. So the fork is no longer necessary.

This has the advantage of installing the pre-compiled .mpy file for these two libraries (whereas `upip` would have installed the source files).

Picoweb is more complicated. There is no `mip` package for it, and likely never will be. You can install it manually by copying the picoweb directory (containing `__init__.py` and `utils.py`) from https://github.com/pfalcon/picoweb, and the "pkg_resources" dependency from `mip`, or you can use `mip`'s http-install feature via:

```
>>> mip.install("pkg_resources")
>>> mip.install("github:pfalcon/picoweb/picoweb/__init__.py", target="/lib/picoweb")
>>> mip.install("github:pfalcon/picoweb/picoweb/utils.py", target="/lib/picoweb")
```

FWIW, I would recommend looking at `microdot` as an alternative to `picoweb`. We're in the process of adding it to `mip`. See https://github.com/miguelgrinberg/microdot/issues/67

**Reply by simonmccaughey** [reply]
This is excellent information, many thanks for taking the time to help out. That all worked perfectly for me, and I will definitely take a look into microdot

**Reply by Guru-Solidz** [reply]
Hi
thank you for the above information, it is very helpful.

however, when logging is installed instead of ulogging, i get an error when i use picoweb
File "/lib/picoweb/__init__.py", line 308, in run
ImportError: no module named 'ulogging'


Could you kindly help let me know how i can get this issue resolved?

Thanks & regards

**Reply by jimmo** [reply]
Add a file named `/lib/ulogging.py` that contains `from logging import *`. Or modify picoweb to use `logging` instead of `ulogging` before copying it to your device.

**Comment by Guru-Solidz** [comment]
Hi,
Thanks for the update.  It works well now.
*_____________________________________________________________*
*Thanks*
*Guruprasad NK | *+91 98718 43210
Founder & CEO*  | **Solidz.IO*

**Comment by Nriver** [comment]
Upip does not work now which is such a big disaster. GitHub is blocked in some country so mip install from GitHub won't work on the device. Everything has to be manually download on a computer with proxy and manually upload them to device now.

What's the story for changing upip to mip?

**Reply by jimmo** [reply]
For micropython-lib packages, mip does not install from GitHub, it's served from micropython.org.

I am aware of very few third-party (i.e. non-micropython-lib) packages that actually worked correctly with PyPI & upip. If you have any examples I'd be keen to see them.

If you want to do on-device installation, the source can be any HTTP server. So you don't have to manually upload them, you just need to be able to proxy to GitHub.

Overall the migration to mip makes a lot more things work and brings a lot more benefits like automatic bytecode compilation.

We're working on adding support for registering third-party packages into micropython-lib such that they can be served from micropython.org too. See recent discussion here: https://github.com/micropython/micropython-lib/issues/578#issuecomment-1321325123

**Comment by bastelbuud** [comment]
hi, i just discovered that upip has been replaced with mip and have some issues as well. Based on this discussions i wanted to upload  a max30102 driver , by doing the following 
mip.install("github:n-elia/MAX30102-MicroPython-driver/max30102/max30102.py", target="/lib/max30102")
but I get the following errors

```
Downloading github:n-elia/MAX30102-MicroPython-driver/max30102/max30102.py to /lib/max30102
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mip/__init__.py", line 1, in install
  File "mip/__init__.py", line 1, in _install_package
  File "mip/__init__.py", line 1, in _download_file
  File "urequests.py", line 180, in get
  File "urequests.py", line 76, in request
```

could you please help in getting this fixed?

**Reply by n-elia** [reply]
Hi bastelbuud,

I'm the maintainer of that package. I created the `package.json` document, therefore the repository is now `mip`-compliant and you're able to install the package including this line into your firmware:

``` python
mip.install("github:n-elia/MAX30102-MicroPython-driver")
```
According to the documentation, my package needs a `package.json` because it is composed of multiple files. When you run the line written above, `mip` looks for that .json file into the repository root, and there it can understand which files must be downloaded.

Thanks for triggering this update!
Nicola

**Reply by n-elia** [reply]
@jimmo: I read that you were looking for micropython-specific modules hosted on PyPI index. This could be the case: my [max30102 driver](https://github.com/n-elia/MAX30102-MicroPython-driver) is an example of mycropython-specific module being hosted on PyPI and successfully installed many times with upip. The repository includes also a CD pipeline for deploying to PyPI if you want to take a look, and I'm available to answer any question.

**Reply by jimmo** [reply]
@bastelbuud As best I can tell, this suggests that maybe your board wasn't connected to the internet? Or at least unable to resolve github.com?

@n-elia Thank you for updating your driver, that's fantastic!  (Although from what I can tell from the git history, it should have worked previously with what @bastelbuud was doing).

Also thanks for providing the example, it's one of the very few MicroPython packages I've seen published to PyPI that does all the required bits for upip. As I wrote in https://github.com/micropython/micropython-lib/issues/578#issuecomment-1321325123 we're going to make it possible to add third-party libraries directly to `mip` so I will reach out to you when we do that (soon!).

**Comment by nathan-stewart** [comment]
Is this the appropriate place to say "Please put it back?"  Mip doesn't have anywhere remotely near the library that upip did. I did get a package I was looking for installed six months ago but I've since forgotten and it doesn't seem to have gotten any better.

**Reply by Josverl** [reply]
Could you help by mentioning some of the modules you used to install using upip?

**Reply by jimmo** [reply]
>  Mip doesn't have anywhere remotely near the library that upip did.

@nathan-stewart In some sense this is deliberate -- the fact that you could install "any" PyPI package from upip was a source of confusion. mip will only install things that are designed for micropython.

If people really want to be able to install arbitrary packages from PyPI then there are three options:
1. Add a separate `pip` command to mpremote (and maybe also implement a `pip` module in micropython-lib for people who want to use it on-device, using the old `upip` as a starting point -- I believe there's already a PR for this).
2. In many cases, it's simple enough to just use the regular CPython `pip` to install the package locally and then copy the files over (e.g. with `mpremote cp -r`)
3. Or for most PyPI packages you can follow the link to their GitHub page and grab the files from there.
